### PR TITLE
Add getBlockEntryByHash rpc call

### PR DESCRIPTION
**Pull request by liuchengxu**
getBlockEntryByHash gets the block header and transaction entries in a
block via its hash.

### CODE REVIEW

File: blockchain/reactor.go [hunk 1]
```diff
@@ -345,6 +348,24 @@ func (bcR *BlockchainReactor) getBestBlockHash() *bc.Hash {
...
+	hash.UnmarshalText([]byte(strHash))
+	block, err := bcr.chain.GetBlockByHash(&hash)
+	if err != nil {
+		log.WithField("error", err).Error("Fail to get block by hash")
+	}
+	bcBlock := legacy.MapBlock(block)
+	header, _ := stdjson.MarshalIndent(bcBlock.BlockHeader, "", "  ")
+	buf.WriteString(string(header))
+	for _, transaction := range bcBlock.Transactions {
+		tx, _ := stdjson.MarshalIndent(transaction, "", "  ")
```
Comment (Paladz) [inline blockchain/reactor.go]: for id, e := range transaction.Entries {
	switch e := e.(type) {
		case *bc.Issuance:
		case *bc.Nonce:
		case *bc.Spend:
		case *bc.Retirement:
		case *bc.Output:
	}
}

循环内部这样读取每一个input和output

File: blockchain/reactor.go [hunk 2]
```diff
@@ -345,6 +348,24 @@ func (bcR *BlockchainReactor) getBestBlockHash() *bc.Hash {
 	return bcR.chain.BestBlockHash()
 }
 
+func (bcr *BlockchainReactor) getBlockEntryByHash(strHash string) string {
+	var buf bytes.Buffer
+	hash := bc.Hash{}
+	hash.UnmarshalText([]byte(strHash))
+	block, err := bcr.chain.GetBlockByHash(&hash)
+	if err != nil {
+		log.WithField("error", err).Error("Fail to get block by hash")
```
Comment (Paladz) [inline blockchain/reactor.go]: add rturn if failed

File: blockchain/reactor.go [hunk 3]
```diff
@@ -365,6 +366,50 @@ func (bcr *BlockchainReactor) getBlockHeaderByHash(strHash string) string {
 	return buf.String()
 }
 
+func (bcr *BlockchainReactor) getBlockEntryByHash(strHash string) string {
+	var buf bytes.Buffer
+	hash := bc.Hash{}
+	if err := hash.UnmarshalText([]byte(strHash)); err != nil {
+		log.WithField("error", err).Error("Error occurs when transforming string hash to hash struct")
```
Comment (Paladz) [inline blockchain/reactor.go]: return something?

File: blockchain/reactor.go [hunk 4]
```diff
@@ -365,6 +366,50 @@ func (bcr *BlockchainReactor) getBlockHeaderByHash(strHash string) string {
...
+	}
+	block, err := bcr.chain.GetBlockByHash(&hash)
+	if err != nil {
+		log.WithField("error", err).Error("Fail to get block by hash")
+		return ""
```
Comment (Paladz) [inline blockchain/reactor.go]: return err string?

File: blockchain/reactor.go [hunk 5]
```diff
@@ -365,6 +366,50 @@ func (bcr *BlockchainReactor) getBlockHeaderByHash(strHash string) string {
 	return buf.String()
 }
 
+func (bcr *BlockchainReactor) getBlockEntryByHash(strHash string) string {
```
Comment (Paladz) [inline blockchain/reactor.go]: getBlockByHash ？

File: blockchain/reactor.go (hunk 4 above)
Comment (liuchengxu) [reply blockchain/reactor.go]: You mean `return "err"`?

File: blockchain/reactor.go (hunk 5 above)
Comment (liuchengxu) [reply blockchain/reactor.go]: 👌

File: blockchain/reactor.go [hunk 6]
```diff
@@ -365,6 +366,50 @@ func (bcr *BlockchainReactor) getBlockHeaderByHash(strHash string) string {
...
+		log.WithField("error", err).Error("Error occurs when transforming string hash to hash struct")
+	}
+	block, err := bcr.chain.GetBlockByHash(&hash)
+	if err != nil {
+		log.WithField("error", err).Error("Fail to get block by hash")
+		return ""
+	}
+	bcBlock := legacy.MapBlock(block)
+	header, _ := stdjson.MarshalIndent(bcBlock.BlockHeader, "", "  ")
+	// FIXME: smarter pretty-print
```
Comment (Paladz) [inline blockchain/reactor.go]: GetBlockByHash struct {
    BlockHeader *BlockHeader "json: xxxx"
    Inputs *[]tx.Entries "json: xxxx"
    Outputs  *[]tx.Entries "json: xxxx"
}

然后把读出来的参数全部指针传进去， 最后只要stdjson.Marshal 一次就是标准的json了

Comment (liuchengxu) [reply blockchain/reactor.go]: Done.

File: blockchain/reactor.go [hunk 7]
```diff
@@ -365,6 +366,47 @@ func (bcr *BlockchainReactor) getBlockHeaderByHash(strHash string) string {
...
 }
 
+type GetBlockByHash struct {
+	BlockHeader *bc.BlockHeader `json:"block_header"`
+	Inputs      []bc.Entry      `json:"inputs"`
+	Outputs     []bc.Entry      `json:"outputs"`
+}
+
+func (bcr *BlockchainReactor) getBlockByHash(strHash string) string {
+	hash := bc.Hash{}
```
Comment (Paladz) [inline blockchain/reactor.go]: hash := NewHash()  //bc/hash.go

File: blockchain/reactor.go [hunk 8]
```diff
@@ -365,6 +366,47 @@ func (bcr *BlockchainReactor) getBlockHeaderByHash(strHash string) string {
...
+	if err := hash.UnmarshalText([]byte(strHash)); err != nil {
+		log.WithField("error", err).Error("Error occurs when transforming string hash to hash struct")
+	}
+	legacyBlock, err := bcr.chain.GetBlockByHash(&hash)
+	if err != nil {
+		log.WithField("error", err).Error("Fail to get block by hash")
+		return ""
```
Comment (Paladz) [inline blockchain/reactor.go]: return error string

File: blockchain/reactor.go [hunk 9]
```diff
@@ -365,6 +366,47 @@ func (bcr *BlockchainReactor) getBlockHeaderByHash(strHash string) string {
...
+	}
+	block := &GetBlockByHash{}
```
Comment (Paladz) [inline blockchain/reactor.go]: resp ???

File: blockchain/reactor.go [hunk 10]
```diff
@@ -365,6 +366,47 @@ func (bcr *BlockchainReactor) getBlockHeaderByHash(strHash string) string {
 	return buf.String()
 }
 
+type GetBlockByHash struct {
+	BlockHeader *bc.BlockHeader `json:"block_header"`
+	Inputs      []bc.Entry      `json:"inputs"`
```
Comment (Paladz) [inline blockchain/reactor.go]: +type GetBlockByHash struct {
 	BlockHeader *bc.BlockHeader `json:"block_header"`
        TX struct []{
         Inputs      []bc.Entry      `json:"inputs"`
 	Outputs     []bc.Entry      `json:"outputs"`
       }
 }

File: blockchain/reactor.go [hunk 11]
```diff
@@ -365,6 +366,52 @@ func (bcr *BlockchainReactor) getBlockHeaderByHash(strHash string) string {
 	return buf.String()
 }
 
+type GetBlockByHash struct {
+	BlockHeader  *bc.BlockHeader `json:"block_header"`
+	Transactions struct {
+		tx      []*bc.Tx
+		Inputs  []bc.Entry `json:"inputs"`
```
Comment (Paladz) [inline blockchain/reactor.go]: inputs and output should be inside the transaction

File: blockchain/reactor.go [hunk 12]
```diff
@@ -365,6 +366,52 @@ func (bcr *BlockchainReactor) getBlockHeaderByHash(strHash string) string {
...
+func (bcr *BlockchainReactor) getBlockByHash(strHash string) string {
+	hash := bc.Hash{}
+	if err := hash.UnmarshalText([]byte(strHash)); err != nil {
+		log.WithField("error", err).Error("Error occurs when transforming string hash to hash struct")
+		return "Error"
+	}
+	legacyBlock, err := bcr.chain.GetBlockByHash(&hash)
+	if err != nil {
+		log.WithField("error", err).Error("Fail to get block by hash")
+		return "Error"
```
Comment (Paladz) [inline blockchain/reactor.go]: err.string()

**Pull request merged**